### ISSUE TITLE
Variable sized batches

### ISSUE THREAD

**Issue by pmrv** [google/grain#1154]
We are trying to solve the following problem in grain.  Our data for our JAX models requires postprocessing before it can be fed to the models and this postprocessing produces variable sized data.  Because of this we batch and pad multiple input items together to achieve batches of the same size to avoid recompilation.  The postprocessed data is much larger than the original input items.  
Our problem is now that the `Batched` transformation for `DataLoader` in grain only allows to specify a fixed number of input items for a fixed batch size, whereas we would rather ingest a variable number of input items until we can reach a given output batch size with minimal padding.  `Batched` doesn't seem to support this.  Is this something that could be done with the `DataSet` API or do you see another way of mapping the problem onto the `DataLoader`?

**Comment by mhyatt000** [comment]
Not a maintainer, but I have found dataset API to be more flexible. More info about the problem might help determine best solution. 

ie: how is the variable batch size determined? is it purely rng or is there some measurement you can use to batch intelligently

I also suffer from the same problem of too much padding, but in my case the data must all be shuffled together so I accept the additional padding. conversely, it sounds like you are willing to remove the padding at the cost of some randomness in the data order

**Comment by pmrv** [comment]
Thanks for you input!  The data order is not so important to us.  The batch is determined mostly by the memory of the accelerator devices we run on.  The difficulty for us that from the data elements themselves it's tricky to predict the size of the transformed data without running the actual transform, even though the relationship is deterministic.  I'll try and work out a minimal example in python below.

The following code basically reflects the behaviour we want in pure python iterators

```python
N = 10
source = 1.0 * np.random.randint(low=1, high=10, size=N)
def my_complex_transform(data):
    return jnp.arange(data)  # imagine this was something more complex where the shape is not easily known a priori
def variable_batcher(target_size, source):
    batch = []
    for data in source:
        if len(batch) + len(data) <= target_size:
            batch = np.concatenate([batch, data])
        else:
            yield np.pad(batch, (0, target_size - len(batch)), constant_values=-1)
            batch = data
for data in source:
    print(data)
for transformed_data in map(my_complex_transform, source):
    print(transformed_data)
for transformed_batch in variable_batcher(15, map(my_complex_transform, source)):
    print(transformed_batch)
```

with this output

<img width="913" height="601" alt="Image" src="https://github.com/user-attachments/assets/177ae276-7e22-4ac5-9e2c-1cc94884ea6c" />

Reading a bit more in the docs, it seems `FirstFitPackIterDataset` or `ConcatThenSplitIterDataset` may do what we want.  Unless the devs disaprove I'll document the success of that here, seems as good a place as any.

**Comment by pmrv** [comment]
`FirstFitPackIterDataset` seems to get us close.  Since docs are sparse here's an example

```python
import jax.numpy as jnp
import numpy as np
import grain.python as grain
class MyVariableTransform(grain.MapTransform):
    counter = 0
    def map(self, data):
        # return {'range': list(range(data))}  # won't work
        # return {'range': jnp.arange(data)} # won't work
        return {
            'constant': data * np.ones(data),
            'total': data,
        }
ds = grain.MapDataset.source(np.random.randint(1, 10, 20)).apply(MyVariableTransform())
bs = grain.experimental.FirstFitPackIterDataset(ds, length_struct={
    'constant': 20, 
    'total': 5,
}, num_packing_bins=1,)
for data in ds:
    print(data)
for data in bs:
    pprint(data)
```

which yields items like this, including positions/segment_ids for segmented sums, which is nice!

<img width="739" height="159" alt="Image" src="https://github.com/user-attachments/assets/475d3f4f-c78f-4816-99fa-77ce1eb2dac8" />

Note that the underlying data source needs to emit dictionaries of numpy arrays or scalars. I'm not entirely clear on all the options it supports, but I'll ask those in separate issues.
